pallene/to_ir.lua

```diff
@@ -209,60 +209,99 @@ function ToIR:convert_stat(cmds, stat)
         local decls = stat.decls
         local exps = stat.exps
 
+        local is_ipairs = false
+        local iterfn = exps[1]
+        if iterfn._tag == "ast.Exp.CallFunc" and iterfn.exp._tag == "ast.Exp.Var" then
+            local iterfn_var = iterfn.exp.var._name
+            is_ipairs = iterfn_var._tag == "checker.Name.Builtin" and iterfn_var.name == "ipairs"
+        end
+
 
-        if stat.is_ipairs then
+        if is_ipairs then
             -- `ipairs` are desugared down to regular for-loops
             -- ```
-            -- for i, x in ipairs(xs) do
+            -- for i: T1, x: T2 in ipairs(xs) do
             --   <loop body>
             -- end
             -- ```
             -- would get compiled down to:
             -- ```
-            -- local x
-            -- local i_num
-            -- for i_num = 1, #xs do
-            --   i = i_num
-            --   x = xs[i]
+            -- local x_dyn: any
+            -- local i_num: integer = 1
+            -- local limit: integer = #xs
+            -- while true do
+            --   if i_num > limit then
+            --     break
+            --   end
+            --   local i = i_num as T1
+            --   x_dyn = xs[i_num]
+            --   if x_dyn == nil then
+            --     break
+            --   end
+            --   local x = x_dyn as T2 
             --   <loop body>
+            --   i_num = i_num + 1
```

Comment:
> In order for the `xs[i_num]` to produce an any as we want, the xs must have type {any}. Are we already doing this?

Would you say that's the checker's job to make sure the ipair's argument is of type `{any}`?
Or do we do it in `to_ipairs`?

> ps.: By the way, I think I found a small bug with the comment for the generic for-in loop. We should be using a_any as the control variable instead of ctrl.

Whoops, yeah I missed that.

Reply:
I think the checker probably already does it, but I don't remember 100%. Might be worth adding an `assert` call, just in case.